simd_emulation/src/lib.rs

```diff
@@ -0,0 +1,1234 @@
+        let r2 = f(A::get2(a));
+        let r3 = f(A::get3(a));
+        let r4 = f(A::get4(a));
+        let r5 = f(A::get5(a));
+        let r6 = f(A::get6(a));
+        let r7 = f(A::get7(a));
+        let r8 = f(A::get8(a));
+        let r9 = f(A::get9(a));
+        let r10 = f(A::get10(a));
+        let r11 = f(A::get11(a));
+        let r12 = f(A::get12(a));
+        let r13 = f(A::get13(a));
+        let r14 = f(A::get14(a));
+        let r15 = f(A::get15(a));
+        reduce16(
+            g,
+            [
+                r0, r1, r2, r3, r4, r5, r6, r7, r8, r9, r10, r11, r12, r13, r14, r15,
+            ],
+        )
+    }
+
+    // lift a unary operation over a vector and reduce
+    pub fn lift32_v_s<F, G, A, R>(f: F, g: G, a: A::Vec) -> R
+    where
+        F: Fn(A) -> R,
+        G: Fn(usize, R, R) -> R,
+        A: Vector32,
+        R: Copy,
+    {
+        let r0 = f(A::get0(a));
+        let r1 = f(A::get1(a));
+        let r2 = f(A::get2(a));
+        let r3 = f(A::get3(a));
+        let r4 = f(A::get4(a));
+        let r5 = f(A::get5(a));
+        let r6 = f(A::get6(a));
+        let r7 = f(A::get7(a));
+        let r8 = f(A::get8(a));
+        let r9 = f(A::get9(a));
+        let r10 = f(A::get10(a));
+        let r11 = f(A::get11(a));
+        let r12 = f(A::get12(a));
+        let r13 = f(A::get13(a));
+        let r14 = f(A::get14(a));
+        let r15 = f(A::get15(a));
+        let r16 = f(A::get16(a));
+        let r17 = f(A::get17(a));
+        let r18 = f(A::get18(a));
+        let r19 = f(A::get19(a));
+        let r20 = f(A::get20(a));
+        let r21 = f(A::get21(a));
+        let r22 = f(A::get22(a));
+        let r23 = f(A::get23(a));
+        let r24 = f(A::get24(a));
+        let r25 = f(A::get25(a));
+        let r26 = f(A::get26(a));
+        let r27 = f(A::get27(a));
+        let r28 = f(A::get28(a));
+        let r29 = f(A::get29(a));
+        let r30 = f(A::get30(a));
+        let r31 = f(A::get31(a));
+        reduce32(
+            g,
+            [
+                r0, r1, r2, r3, r4, r5, r6, r7, r8, r9, r10, r11, r12, r13, r14, r15, r16, r17,
+                r18, r19, r20, r21, r22, r23, r24, r25, r26, r27, r28, r29, r30, r31,
+            ],
+        )
+    }
+}
+
+mod scalar {
+    // todo: there may be some room for sharing code between the different int sizes/signs?
+
+    pub fn and64(x: u64, y: u64) -> u64 {
+        x & y
+    }
+
+    pub fn andnot64(x: u64, y: u64) -> u64 {
+        x & !y
+    }
+
+    pub fn or64(x: u64, y: u64) -> u64 {
+        x | y
+    }
+
+    pub fn xor64(x: u64, y: u64) -> u64 {
+        x ^ y
+    }
+
+    pub fn cmpeq_u8(x: u8, y: u8) -> u8 {
+        if x == y {
+            0xff
+        } else {
+            0x0
+        }
+    }
+
+    pub fn cmpeq_u16(x: u16, y: u16) -> u16 {
+        if x == y {
+            0xffff
+        } else {
+            0x0
+        }
+    }
+
+    pub fn cmpgt_u8(x: u8, y: u8) -> u8 {
+        if x == y {
```

Comment:
From the `gt` in the name I would expect `>` instead of `==`.

Reply:
oops
Though, on closer inspection, I noticed that the function was dead code and deleted it.
(I used to define more intrinsics - but then I noticed that LLVM was turning them into generic (architecture independent) intrinsics so I did not need to define them after all.)